src/components/demo-tabs.tsx

```diff
@@ -0,0 +1,183 @@
+import React from 'react';
+import { StaticQuery, graphql } from 'gatsby';
+import { ContentPageData } from '../types';
+import AuthorsList from './authors-list';
+import EditLink from './edit-link';
+import { ParsedContent } from '../util/ParsedContent';
+
+const query = graphql`
+  query DemoTabsContent {
+    content: markdownRemark(
+      frontmatter: {
+        title: { eq: "Installing via Package Manager" }
+        section: { eq: "Content" }
+      }
+    ) {
+      id
+      html
+      tableOfContents
+      frontmatter {
+        title
+        description
+      }
+      fields {
+        slug
+        authors
+      }
+      parent {
+        ... on File {
+          relativePath
+        }
+      }
+    }
+  }
+`;
+
+export interface Props {
+  parsedContent?: ParsedContent;
+}
+
+class DemoTabs extends React.Component<Props> {
+  public parsedContent: ParsedContent;
+
+  public parsedFragment: DocumentFragment | null;
+
+  public articleRef: React.RefObject<HTMLElement>;
+
+  public selectRef: React.RefObject<HTMLSelectElement>;
+
+  public optionMap: WeakMap<HTMLOptionElement, HTMLElement>;
+
+  public options: HTMLOptionElement[];
+
+  public sections: HTMLElement[];
+
+  public currentSection: HTMLElement | null;
+
+  public constructor(props: Props) {
+    super(props);
+    this.parsedContent = (props && props.parsedContent) || new ParsedContent();
+    this.articleRef = React.createRef();
+    this.selectRef = React.createRef();
+    this.optionMap = new WeakMap<HTMLOptionElement, HTMLElement>();
+    this.options = [];
+    this.sections = [];
+    this.currentSection = null;
+    this.parsedFragment = null;
+  }
+
+  public componentDidMount() {
+    this.populate();
+  }
+
+  public populate() {
+    if (this.parsedFragment === this.parsedContent.documentFragment) return;
+
+    /* eslint-disable */
+
+    const {
+      parsedContent,
+      parsedFragment,
+      articleRef: { current: articleNode },
+      selectRef: { current: selectNode },
+    } = this;
+
+    const splicedNodes = [
+      ...this.options.splice(0, this.options.length),
+      ...this.sections.splice(0, this.sections.length),
+    ];
+
+    this.currentSection = null;
+
+    // @ts-ignore
+    for (const node of splicedNodes) node.remove();
+
+    if (!articleNode || !selectNode) return;
+
+    if (parsedContent.sections) {
+      for (const section of parsedContent.sections) {
+        const clone = section.cloneNode(true) as HTMLElement;
+        clone.removeAttribute('visible');
+        const name = clone.getAttribute('name') || '';
+        const option = selectNode.appendChild(document.createElement('option'));
+        this.optionMap.set(option, clone);
+        option.setAttribute('value', (option.innerText = name));
+        this.sections.push(clone);
+        this.options.push(option);
+        const heading = /** @type {HTMLHeadingElement} */ clone.querySelector(
```

Comment:
Why `/* @type */` instead of `const header: HTMLHeadingElement`?

Also, let's use JSX-y convention, since we're in React world here. You should be able to return an array of constructed `<option>` JSX elements in this method and let the renderer do the rest.

Reply:
Nice catch… but that will likely go away anyways.

I'm doing the following:

1. Use a ParsedContent instance to create a document fragment and parse nodes in a document fragment, exposing an array of `<section name="…">` nodes.

2. Populate a WeakMap of each  `<section>` mapped via a respective `<option>` node and retain an array for each through the life time of the component.

The motivating aspect in scalability of this pattern is to avoid garbage-collection related pitfalls.

Once we have clean parsing it should all be JSX, I just need my head in the right space to cross that divide which I am still not comfortable to do at this point.

So a question for you, is there a Reacty parallel to `querySelector…` and `remove()`? I'm sure on your end you have more insights on this.